### PR TITLE
feat(KFLUXVNGD-1114): detect image digest bumps in operator changelog comment

### PR DESCRIPTION
**Pull request by kelchen123**
Resolve digest-pinned upstream components via OCI labels and include them in the existing service bump and commit summary pipeline. Degrade when digest changes are detected but registry resolution produces no bumps.

Assisted-by: Cursor

### CODE REVIEW

File: infra-tools/internal/changelog/images.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+//
+// Algorithm (single forward pass):
+//  1. A '-' line matching digest: sha256:... sets pendingOld.
+//  2. A '+' line matching digest: sha256:... sets pendingNew.
+//  3. A context ' ' line matching \bname: (not newName:) closes the pending pair
+//     and emits an ImageDigestChange.
+//  4. A context ' ' line that starts a new YAML list item ("- ") resets the
+//     pending state — we crossed an unchanged image entry.
+//
+// This ordering is correct for the real kustomization files: digest appears
```
Comment (yftacherzog) [inline infra-tools/internal/changelog/images.go]: We cannot (and should not) rely on the order. This one for example will not be processed: https://github.com/konflux-ci/konflux-ci/blob/main/operator/upstream-kustomizations/ui/dex/kustomization.yml

We need something more robust.

**Review by yftacherzog — CHANGES_REQUESTED** [review]
See my comment about yaml fields ordering.

File: infra-tools/internal/changelog/images_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	g.Expect(err).To(HaveOccurred())
+}
+
+func TestInspectLabels_NonexistentDigest(t *testing.T) {
+	g := NewWithT(t)
+	r := changelog.NewRegistryInspector()
+	digest := "sha256:" + strings.Repeat("0", 64)
+	_, err := r.InspectLabels(context.Background(), "quay.io/konflux-ci/namespace-lister@"+digest)
+	g.Expect(err).To(HaveOccurred())
+}
```
[the remark this answers is not in the document]
Comment (yftacherzog) [reply infra-tools/internal/changelog/images_test.go]: @kelchen123 not sure we need this test for coverage. Anyhow, I don't think we should introduce network dependency in unit tests. If we want the test, then we should make the dependency injectable.

File: infra-tools/internal/changelog/images.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+// InspectLabels fetches the config labels for the given image reference
+// (e.g. "quay.io/konflux-ci/namespace-lister@sha256:...").
+// The call is retried up to three times with exponential backoff.
+func (r *ocRegistryInspector) InspectLabels(ctx context.Context, imageRef string) (map[string]string, error) {
+	var labels map[string]string
+	err := retryDo(ctx, 3, func() error {
+		ref, err := name.ParseReference(imageRef)
+		if err != nil {
+			return fmt.Errorf("parsing image reference %s: %w", imageRef, err)
+		}
```
[the remark this answers is not in the document]
Comment (yftacherzog) [reply infra-tools/internal/changelog/images.go]: @kelchen123 I think this is valid. We should probably move the parsing out of what is being retried.

**Review by yftacherzog — COMMENTED** [review]
@kelchen123 direction is good! See my comments and also Qodo's comments (I did not check if all are relevant) for fixing the remaining issues.

**Review by yftacherzog — APPROVED** [review]

File: infra-tools/cmd/changelog-generator/main.go [hunk 4]
```diff
@@ -256,6 +267,65 @@ func formatCompare(oldRef, newRef string, bumps []bumpWithCommits) string {
...
+			continue
+		}
+		newSHA, err := resolveDigestToSHA(ctx, inspector, c.ImageName, c.NewDigest)
+		if err != nil {
+			slog.Warn("resolving new digest to git SHA", "image", c.ImageName, "digest", c.NewDigest[:19], "err", err)
+			continue
+		}
+		bumps = append(bumps, changelog.ServiceBump{Owner: owner, Repo: repo, OldSHA: oldSHA, NewSHA: newSHA})
+	}
+	return bumps
```
[the remark this answers is not in the document]
Comment (gbenhaim) [reply infra-tools/cmd/changelog-generator/main.go:300]: @kelchen123 is this a real issue?

**Comment by gbenhaim** [comment]
@kelchen123 can you please check the review comments from Qodo? if something isn't relevant you can close it.

File: infra-tools/cmd/changelog-generator/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kelchen123) [reply infra-tools/cmd/changelog-generator/main.go:300]: @gbenhaim Not a real issue for this PR anymore.

`resolveImageBumps` already dedupes by `owner/repo` before registry lookups (added in the follow-up commit):

```go
key := owner + "/" + repo
if seen[key] {
    continue
}
seen[key] = true
```

So duplicate image entries for the same repo won't produce duplicate bumps or redundant `FetchServiceCommits` calls. Closing this thread.

File: infra-tools/cmd/changelog-generator/main.go [hunk 5]
```diff
@@ -256,6 +267,65 @@ func formatCompare(oldRef, newRef string, bumps []bumpWithCommits) string {
...
+
+// extractOwnerRepoFromImageName splits a container image reference of the form
+// "registry/owner/repo[:tag]" or "registry/owner/repo@digest" and returns
+// (owner, repo). Returns ("","") if the image name has fewer than three
+// slash-separated components.
+func extractOwnerRepoFromImageName(imageName string) (owner, repo string) {
+	// Strip tag or digest suffix before splitting on '/'.
+	base := strings.SplitN(imageName, ":", 2)[0]
+	base = strings.SplitN(base, "@", 2)[0]
+	parts := strings.Split(base, "/")
```
[the remark this answers is not in the document]
Comment (kelchen123) [reply infra-tools/cmd/changelog-generator/main.go:329]: Not relevant for our inputs. Konflux upstream images are always `quay.io/konflux-ci/<repo>` (no registry port). Splitting on the first `:` only strips an optional tag, which we don't use on digest-pinned refs anyway (`image@sha256:...`). Closing.

File: infra-tools/cmd/changelog-generator/main.go [hunk 6]
```diff
@@ -257,6 +268,71 @@ func formatCompare(oldRef, newRef string, bumps []bumpWithCommits) string {
...
+	labels, err := inspector.InspectLabels(ctx, imageName+"@"+digest)
+	if err != nil {
+		return "", fmt.Errorf("inspecting %s@%s: %w", imageName, digest[:19], err)
+	}
+	// org.opencontainers.image.revision is the standard OCI label for the git SHA.
+	// vcs-ref is an older convention used as a fallback.
+	for _, key := range []string{"org.opencontainers.image.revision", "vcs-ref"} {
+		if sha := labels[key]; sha != "" {
+			return sha, nil
+		}
```
[the remark this answers is not in the document]
Comment (kelchen123) [reply infra-tools/cmd/changelog-generator/main.go:316]: Closing as out of scope for this PR. Konflux-built images set `org.opencontainers.image.revision` to a full git SHA. If a label were malformed, `FetchServiceCommits` would fail for that service and we'd already surface `_Commit details unavailable._` rather than invent a bump. Happy to add a 40-hex check in a follow-up if we want belt-and-suspenders.

File: infra-tools/cmd/changelog-generator/main.go [hunk 7]
```diff
@@ -257,6 +268,71 @@ func formatCompare(oldRef, newRef string, bumps []bumpWithCommits) string {
...
+			slog.Warn("resolving old digest to git SHA", "image", c.ImageName, "digest", c.OldDigest[:19], "err", err)
+			continue
+		}
+		newSHA, err := resolveDigestToSHA(ctx, inspector, c.ImageName, c.NewDigest)
+		if err != nil {
+			slog.Warn("resolving new digest to git SHA", "image", c.ImageName, "digest", c.NewDigest[:19], "err", err)
+			continue
+		}
+		bumps = append(bumps, changelog.ServiceBump{Owner: owner, Repo: repo, OldSHA: oldSHA, NewSHA: newSHA})
+	}
```
[the remark this answers is not in the document]
Comment (kelchen123) [reply infra-tools/cmd/changelog-generator/main.go:299]: Theoretical edge case (two digests, same revision label — e.g. rebuild). Uncommon for our operator bumps; digest-level equality is already filtered in `digestChangeFromBlock`. Closing for this PR; easy one-line `oldSHA == newSHA` skip if we see it in practice.

**Comment by kelchen123** [comment]
@gbenhaim Triaged the open Qodo threads:

| Finding | Disposition |
|---------|-------------|
| Image bumps not deduped | Already fixed in-tree (`seen` map in `resolveImageBumps`) — not a real issue |
| Port registry parsing (`registry:5000/...`) | N/A for Konflux (`quay.io/konflux-ci/...` only) — closed |
| Unvalidated revision label | Out of scope; Konflux labels are full SHAs; bad labels already degrade per-service — closed |
| Same-SHA digest bump | Theoretical rebuild edge case — closed; easy follow-up if needed |

Already-resolved earlier: registry SSRF concern, retry/parse, parser state leak, networked unit test, `newName` handling, field-order/block parser, digest-first adjacent-entry heuristic.
